datastore/datastore.go

```diff
@@ -142,3 +144,36 @@ func (ds *Datastore) GetGetPayloadResponse(slot uint64, proposerPubkey, blockHas
 		Data:    executionPayload,
 	}, nil
 }
+
+// SetBlockBuilderStatusByCollateralID modifies the status of a set of builders
+// who have matching collateral IDs. This function continues to try change
+// the statuses, even if some fail. The returned slice contains all the errors
+// encountered.
+func (ds *Datastore) SetBlockBuilderStatusByCollateralID(builderPubkey string, status common.BuilderStatus) []error {
+	// Fetch builder collateral_id.
+	builder, err := ds.db.GetBlockBuilderByPubkey(builderPubkey)
+	if err != nil {
+		err = fmt.Errorf("unable to get builder from database: %v", err)
+		ds.log.Error(err.Error())
+		return []error{err}
+	}
+
+	// Fetch all builder pubkeys using the collateral_id.
+	builderPubkeys, err := ds.db.GetBlockBuilderPubkeysByCollateralID(builder.CollateralID)
+	if err != nil {
+		err = fmt.Errorf("unable to get builder pubkeys by collateral id: %v", err)
+		ds.log.Error(err.Error())
+		return []error{err}
+	}
+
+	var errs []error
+	for _, pubkey := range builderPubkeys {
+		err := ds.db.SetBlockBuilderStatus(pubkey, status)
+		if err != nil {
+			err = fmt.Errorf("failed to set block builder: %v status in db: %v", pubkey, err)
+			ds.log.Error(err.Error())
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
```

Comment:
Since Redis is no longer involved should this go in `database.go`? Also, would it be possible to do a one-shot SQL update (instead of 1 + 1 reads and `n` writes)? That may be cleaner and more performant. A single error object would suffice.

Reply:
nice, good points! here is a commit that implements this. https://github.com/michaelneuder/mev-boost-relay/pull/3/commits/8425fbcb28b64bece655f2025b63f53128d04307

the tl;dr; is we can bundle this all into a single `SetBuilderStatus` function, which checks if the collateral ID of a builder is set. if it is, then update all the statuses in once DB update call. if it is not, then just update the single builder pubkey. this is a nice cleanup i think!

Reply:
nice!